### PR TITLE
[GPU][Codegen] Choose VDMFMA intrinsics for 16bit input skinny gemms

### PR DESCRIPTION
**Pull request by efric**
Adds 16 bit VDMFMA candidates to the TileAndFuse intrinsic selection for skinny gemms (5 < M < 9) for CDNA3.

Performance comparison:

CDNA3:
```
  shape                     vdmfma      baseline    %
  ------------------------  ----------  ----------  -------
  f16_8x13312x16384        189 us      206 us      +8.3%
  f16_8x13312x8192         117 us      116 us      -
  f16_8x2304x16384         133 us      138 us      +3.6%
  f16_8x2304x8192          103 us      110 us      +6.4%
  f16_8x6656x16384         127 us      130 us      +2.3%
  f16_8x6656x8192          102 us      109 us      +6.4%
```

VDMFMAs use sparse MFMA instructions, which have twice the depth as dense MFMA variants. Because of this, their compute phase is noticeably shorter. On VDMFMA selection, we double the K reduction tile so the compute phase is long enough to hide global memory latency effectively in software pipelining. This brings the performance against the baseline in line with the above figures. 

As part of https://github.com/iree-org/iree/issues/22863

### CODE REVIEW

File: compiler/src/iree/compiler/Codegen/Dialect/GPU/TargetUtils/ConfigUtils.cpp [hunk 1]
```diff
@@ -348,6 +349,21 @@ static std::optional<GPUMMASchedule> getMmaScheduleFromProblemAndTarget(
...
+
+        // VDMFMAs use the sparse trick (smfmac) and are efficient for skinny
+        // GEMMS (M=8).
+        for (VirtualMMAIntrinsic vi : mma.getVirtualIntrinsics()) {
+          if (!isVDMFMAIntrinsic(vi)) {
+            continue;
+          }
+          auto vmma = VirtualMMAAttr::get(ctx, vi);
+          auto [vm, vn, vk] = vmma.getMNKShape();
+          if (problem.mSizes.empty() || llvm::product_of(problem.mSizes) > vm) {
```
Comment (Muzammiluddin-Syed-ECE) [inline compiler/src/iree/compiler/Codegen/Dialect/GPU/TargetUtils/ConfigUtils.cpp]: nit: `problem.mSizes.empty()` sounds like something we either assume is false or error / assert on.

File: compiler/src/iree/compiler/Codegen/Dialect/GPU/TargetUtils/ConfigUtils.cpp [hunk 2]
```diff
@@ -916,6 +932,17 @@ getMatmulOrIGEMMLoweringConfigAndWorkgroupSize(
...
+    // VDMFMA (smfmac sparse trick) compresses the compute phase to roughly half
+    // the cycles of an equivalent MFMA sequence. In a software-pipelined loop
+    // this can cause the compute to finish before the next tile's loads arrive,
+    // turning latency-hiding overlap into vmcnt stalls. Double the K tile count
+    // to restore the compute-to-memory ratio.
+    if (auto vmma = dyn_cast<IREE::GPU::VirtualMMAAttr>(schedule->mmaKind)) {
+      if (isVDMFMAIntrinsic(vmma.getIntrinsic())) {
+        reductionTileSizes[kDim] *= 2;
+      }
+    }
```
Comment (Muzammiluddin-Syed-ECE) [inline compiler/src/iree/compiler/Codegen/Dialect/GPU/TargetUtils/ConfigUtils.cpp:945]: nit: unless I'm misunderstanding this seems like a performance-related change that affects all architectures that support sparsity, not just CDNA4. This feels like something that would be nice to have in its own PR with data that shows an improvement in performance thanks to better latency hiding.

**Review by Muzammiluddin-Syed-ECE — APPROVED** [review]
LGTM % nits

File: compiler/src/iree/compiler/Codegen/Dialect/GPU/TargetUtils/ConfigUtils.cpp (hunk 2 above)
Comment (efric) [reply compiler/src/iree/compiler/Codegen/Dialect/GPU/TargetUtils/ConfigUtils.cpp:945]: VDMFMA is the only user of sparse intrinsics; the effects for CDNA3 variants are also included in the PR description

File: compiler/src/iree/compiler/Codegen/LLVMGPU/test/ROCDL/config_tile_and_fuse.mlir [hunk 3]
```diff
@@ -1173,3 +1173,86 @@ func.func @gemm_with_dps_init_producer(
...
+}
+
+// CHECK-LABEL: func.func @skinny_gemm_m8_f16(
+//  CHECK-SAME:   #iree_codegen.translation_info<pipeline = #iree_gpu.pipeline<TileAndFuse>
+//       CHECK:   linalg.generic {{.*}}lowering_config = #iree_gpu.lowering_config
+//  CHECK-SAME:     mma_kind = #iree_gpu.virtual_mma_layout<VDMFMA_F32_8x16x64x2_F16>
+
+// -----
+
+// Skinny GEMM (M=8, BF16): should select VDMFMA_F32_8x16x64x2_BF16.
```
Comment (krzysz00) [inline compiler/src/iree/compiler/Codegen/LLVMGPU/test/ROCDL/config_tile_and_fuse.mlir:1236]: ... what about M=6? M=3? M=1?

I figure there's a cutoff point where padding in extra zeroes is fine

Comment (efric) [reply compiler/src/iree/compiler/Codegen/LLVMGPU/test/ROCDL/config_tile_and_fuse.mlir:1236]: Well, `m <= 4` goes down matvec (which we should ... it's like 3% better at m=4 right now, and I expect that this gap will increase after the matvec improvements land; at which point we can reevaluate the matvec threshold). `5 < M < 7` will be padded to sparse trick; I added a lit test to show this explicitly.

**Comment by efric** [comment]
Bringing this back now that I fixed all upstream corner cases that CI caught; https://github.com/llvm/llvm-project/pull/195977 https://github.com/llvm/llvm-project/pull/193541 . Support for other architectures to follow (the register layouts differ).

**Comment by kuhar** [comment]
> Bringing this back now that I fixed all upstream corner cases that CI caught; [llvm/llvm-project#195977](https://github.com/llvm/llvm-project/pull/195977) [llvm/llvm-project#193541](https://github.com/llvm/llvm-project/pull/193541) . Support for other architectures to follow (the register layouts differ).

Can you specify in the PR description which targets will be supported once this lands and what improvement to expect?

**Comment by kuhar** [comment]
Also link to your issue / discussion

File: compiler/src/iree/compiler/Codegen/Dialect/GPU/TargetUtils/ConfigUtils.cpp [hunk 4]
```diff
@@ -342,6 +343,34 @@ static std::optional<GPUMMASchedule> getMmaScheduleFromProblemAndTarget(
...
+          auto vmma = VirtualMMAAttr::get(ctx, vi);
+          auto [vm, vn, vk] = vmma.getMNKShape();
+          if (llvm::product_of(problem.mSizes) > vm) {
+            continue;
+          }
+
+          const int64_t vdmfmaKTileSize = vmma.getIntrinsicsK() * vk;
+          const int64_t problemKSize = llvm::product_of(problem.kSizes);
+          if (problemKSize < vdmfmaKTileSize ||
+              problemKSize % vdmfmaKTileSize != 0) {
```
Comment (kuhar) [inline compiler/src/iree/compiler/Codegen/Dialect/GPU/TargetUtils/ConfigUtils.cpp]: the first condition is redundant, no?

**Comment by efric** [comment]
> Can you specify in the PR description which targets will be supported once this lands and what improvement to expect?

This PR only enables e2e support for gfx942. The expected performance differences are in the PR description.

**Review by kuhar — APPROVED** [review]
